pkg/controller/eks/nodegroup/controller.go

```diff
@@ -0,0 +1,239 @@
+	if err := c.kube.Get(ctx, n, s); err != nil {
+		return nil, errors.Wrap(err, errGetProviderSecret)
+	}
+
+	eksClient, stsClient, err := c.newClientFn(ctx, s.Data[p.Spec.CredentialsSecretRef.Key], p.Spec.Region, awsclients.UseProviderSecret)
+	return &external{client: eksClient, sts: stsClient, kube: c.kube}, errors.Wrap(err, errCreateEKSClient)
+}
+
+type external struct {
+	client eks.Client
+	sts    eks.STSClient
+	kube   client.Client
+}
+
+func (e *external) Observe(ctx context.Context, mg resource.Managed) (managed.ExternalObservation, error) {
+	cr, ok := mg.(*v1alpha1.NodeGroup)
+	if !ok {
+		return managed.ExternalObservation{}, errors.New(errNotEKSNodeGroup)
+	}
+
+	rsp, err := e.client.DescribeNodegroupRequest(&awseks.DescribeNodegroupInput{NodegroupName: aws.String(meta.GetExternalName(cr)), ClusterName: cr.Spec.ForProvider.ClusterName}).Send(ctx)
+	if err != nil {
+		return managed.ExternalObservation{}, errors.Wrap(resource.Ignore(eks.IsErrorNotFound, err), errDescribeFailed)
+	}
+
+	current := cr.Spec.ForProvider.DeepCopy()
+	eks.LateInitializeNodeGroup(&cr.Spec.ForProvider, rsp.Nodegroup)
+	if !reflect.DeepEqual(current, &cr.Spec.ForProvider) {
+		if err := e.kube.Update(ctx, cr); err != nil {
+			return managed.ExternalObservation{}, errors.Wrap(err, errKubeUpdateFailed)
+		}
+	}
+
+	cr.Status.AtProvider = eks.GenerateNodeGroupObservation(rsp.Nodegroup)
+	switch cr.Status.AtProvider.Status {
+	case v1alpha1.NodeGroupStatusActive:
+		cr.Status.SetConditions(runtimev1alpha1.Available())
+		resource.SetBindable(cr)
+	case v1alpha1.NodeGroupStatusCreating:
+		cr.Status.SetConditions(runtimev1alpha1.Creating())
+	case v1alpha1.NodeGroupStatusDeleting:
+		cr.Status.SetConditions(runtimev1alpha1.Deleting())
+	default:
+		cr.Status.SetConditions(runtimev1alpha1.Unavailable())
+	}
+
+	return managed.ExternalObservation{
+		ResourceExists:   true,
+		ResourceUpToDate: eks.IsNodeGroupUpToDate(&cr.Spec.ForProvider, rsp.Nodegroup),
+	}, nil
+}
+
+func (e *external) Create(ctx context.Context, mg resource.Managed) (managed.ExternalCreation, error) {
+	cr, ok := mg.(*v1alpha1.NodeGroup)
+	if !ok {
+		return managed.ExternalCreation{}, errors.New(errNotEKSNodeGroup)
+	}
+	cr.SetConditions(runtimev1alpha1.Creating())
+	if cr.Status.AtProvider.Status == v1alpha1.NodeGroupStatusCreating {
+		return managed.ExternalCreation{}, nil
+	}
+	_, err := e.client.CreateNodegroupRequest(eks.GenerateCreateNodeGroupInput(meta.GetExternalName(cr), &cr.Spec.ForProvider)).Send(ctx)
+	// NOTE(hasheddan): we are ignoring invalid request errors here so that we
+	// do not thrash while waiting for a referenced cluster to become ready.
+	// Node group will report that a successful creation request was made so
+	// this can be somewhat misleading.
+	return managed.ExternalCreation{}, errors.Wrap(resource.Ignore(eks.IsErrorInvalidRequest, err), errCreateFailed)
```

Comment:
```suggestion
	return managed.ExternalCreation{}, errors.Wrap(err, errCreateFailed)
```

I think my comment got lost between reviews but that was the reason I hadn't approved the PR in the first pass.

The reconciliation doesn't start until all referenced resources are ready so I wouldn't expect this to result in error in that case. If the user supplied a raw cluster name and it's not ready, I think it should be fine to return error until it works. However, if we ignore all invalid requests then this will cause cases where the creation doesn't work but there is also no error. So, I don't think we should ignore the invalid request errors here since that encapsulates many errors that we don't know.

Reply:
I see your point and believe that the comment I have there reflects the same sentiment. However, in practice, I think it will be common to create a node group at the same time as the cluster, and returning an error here can cause repeated calls to AWS over an extended period of time while the cluster becomes ready. This can lead to hitting rate limits, which could disrupt reconciliation of other resources as well. Since this is an alpha resource and until we have back off built into Crossplane runtime I think doing this is the right call.

Reply:
Maybe I’m missing something here. Is there anything different here than other resources? For example, Subnet does not call create until the vpc it references is ready and if raw vpc id is given instead of reference, it just calls Create every 30 seconds and returns error if it’s not ready. Or NodePool of GKE.

Reply:
@muvaf nope! You are totally correct, I misinterpreted the first comment and also forgot that we do a `shortWait` instead of returning the error in the reconcile requeue here: https://github.com/crossplane/crossplane-runtime/blob/9a9a434f732111c340d5bdd7ecccb486a025ec2c/pkg/reconciler/managed/reconciler.go#L631

Thanks for pointing this out and apologies for me missing the point originally! Will update now :)

Reply:
@muvaf fixed, I will make sure to clean up commit history before merge here :+1: